app/controllers/checkout_controller.rb

```diff
@@ -21,47 +22,56 @@ def show
 
   def update
     case step
-    when :address  then update_address
+    when :address then update_address
     when :delivery then update_delivery
-    when :payment  then update_payment
+    when :payment then update_payment
     when :confirm  then update_confirm
     when :complete then update_complete
     end
     redirect_to_valid_step
   end
 
+  def show_confirm
+    return jump_to(valid_step) if nil_or_invalid?(current_user.orders.last.credit_card)
+  end
+
+  def show_complete
+    return jump_to(valid_step) if nil_or_invalid?(current_user.orders.last.credit_card)
+    @shipping_address = @order.addresses.find_by_address_type(:shipping)
+    @order.deliver
+  end
+
+  private
+
   def show_address
-    @order.update(user: current_user) if @order.user_id.nil?
-    @billing_address = current_user.addresses.find_by_address_type(:billing)
-    @shipping_address = current_user.addresses.find_by_address_type(:shipping)
-    
-    @billing_address ||= @order.addresses.find_by_address_type(:billing)
-    @shipping_address ||= @order.addresses.find_by_address_type(:shipping)
+    @address_service = CheckoutAddressService.new(@order, nil, current_user)
   end
 
   def show_delivery
-    return jump_to(valid_step) unless address_valid?(@order)
-    @deliveries = Delivery.all
+    @delivery_service = CheckoutDeliveryService.new(@order)
+    jump_to(valid_step) unless @delivery_service.show_delivery?
   end
 
   def show_payment
-    return jump_to(valid_step) if @order.delivery.nil?
-    @credit_card = CreditCard.find_by(order_id: current_user.orders.first.id)
-    @credit_card ||= @order.credit_card
-    @credit_card = CreditCard.new if nil_or_invalid?(@credit_card)
+    @payment_service = CheckoutPaymentService.new(@order, nil, current_user)
+    jump_to(valid_step) unless @payment_service.show_payment?
   end
 
-  def show_confirm
-    return jump_to(valid_step) if nil_or_invalid?(current_user.orders.last.credit_card)
+  def update_address
+    @address_service = CheckoutAddressService.new(@order, params)
```

Comment:
Service object should have one public method (usually `call`, but can be `update`, `perform` etc.)
And your `update_address` method should look like this:
```
@address_service = CheckoutAddressService.new(@order, nil, current_user).call
return if @address_service.billing_address.valid? && @address_service.shipping_address.valid?
render_wizard
```
Please check more examples at:
https://medium.freecodecamp.org/service-objects-explained-simply-for-ruby-on-rails-5-a8cc42a5441f
https://hackernoon.com/service-objects-in-ruby-on-rails-and-you-79ca8a1c946e